### DISCUSSION TITLE
trusted node?

### DISCUSSION THREAD

**Discussion by githubques** [markqvist/Reticulum#81] [Nomad Network]
Have 2 cpus setup to specs with T-Beams and the latest Nomadnet & RNode updates.

Receiving this message:

"[Warning] Could not autoselect a propagation node! LXMF propogation will not be available until a trusted node announces on the network."

Have not found in the discussions or the manual pdf details about Trusted Nodes or how to fix this issue. Closest I've gotten is a previous post mentioning Trusted Nodes: 

"Nomad Network will automatically use the nearest trusted node as a propagation node. Or you can manually specify a specific node to use in the program."
- https://github.com/markqvist/Reticulum/discussions/57

Tried to manually specify a node, but haven't had luck doing that either. Please advise.

**Comment by markqvist** [comment]
I realise that in this day and age of all the answers existing primarily *online*, the offline documentation approach that `nomadnet` is taking is a bit weird. It includes a complete guide, inside the program itself, which aims to explain everything there is to know about using it. I'll allow myself to be a bit lazy here and just copy the text from the guide :)

The TL;DR is this: If nobody else on your network is hosting any nodes, you can host one yourself, by just setting `enable_node = yes` in the configuration. Others (and yourself) can then use this node for offline / propagated delivery. The explanation with a little more detail follows here:

If you go to the `[ Guide ]` tab of the program, and select the second topic, "Concepts & Terminology", you will find the following introduction:

>By default, Nomad Network will attempt to deliver messages to a peer directly. This happens by first establishing an encrypted link directly to the peer, and then delivering the message over it.
>
>If the desired peer is not available because it has disconnected from the network, this method will obviously fail. In this case, Nomad Network will attempt to deliver the message to a node, which will store and forward it over the network, for later retrieval by the destination peer. The message is encrypted with an ephemeral key before being transmitted to the network, and is only readable by the intended recipient.
>
>For propagated delivery to work, one or more nodes must be available on the network. If one or more trusted nodes are available, Nomad Network will automatically select the most suitable node to send the message via, but you can also manually specify what node to use.
>
>To select a node manually, go to the `[ Network ]` part of the program, choose the desired node in the `Known Nodes` list, and select the `< Info >` button. In the `Node Info` dialog, you can specify the selected node as the default propagation node.

This requires that someone is hosting one or more nodes on the network. The "Hosting a Node" topic in the guide goes into details, but basically it boils down to:

>To host a node on the network, you must enable it in the configuration file, by setting `*enable_node`* directive to `*yes`*. You should also configure the other node-related parameters such as the node name.

The node will now start showing up in the Announce Stream of other peers, and it can be marked as trusted or as manually set propagation node.

For even more info on how it all works, it is worth reading the `[ Guide ]` section of the program at your leisure, and probably with a cup of coffee or tea :)

**Comment by githubques** [comment]
First off, thanks for the prompt reply and for this overall impressive project. Your mentioning of being able to do video calling (pending someone developing the app) opens up the horizons from what I originally thought this was going to be (decentralized 90's internet with only text and email).

[---REWRITTEN REPLY---] I just wrote out a few paragraphs going into detail about how I've already done everything you suggested a few times over in different versions to no avail and closed it out with a request to explain the hot keys at the bottom of the Nomadnet Terminal like [C-l], [C-x], etc. since I'm an amateur mostly front end programmer with little Python experience.

Had the thought that C meant Ctrl instead of Cmd or "C+[other key]" that I'd been trying and hit "Ctrl+L" and whammo, "Announce Stream" popped up and finally saw the neighboring signal and sent a message that was promptly received on the other end.

Point of saying all this, I recommend adding under "First Run" an explanation of the hot keys and that the first thing to do once setting everything up is to hit "CTRL + L" to access the Announce Stream. Another fix is to change "[C-l]" to "[Ctrl+L]", etc.

When I have time, I'll do what I can to chime in with suggestions on how to improve user facing since for this to grow as you've mapped out with individual node makers selling to the public around them, it's going to need some demystifying to speed up adoption. If I find even more time, I'll learn python to provide some resources to improve the experience for non-techie users.

**Reply by markqvist** [reply]
Thabks so much for all the valuable feedback, that is definitely noted. I am going to make that much more clear in the intro guide 👍  You are definitely right that some things need a fair bit of demystifying at this point :) I appreciate the feedback, and if you get the time to contribute some code, that would be awesome!

Did you know there is also a GUI app called Sideband that interacts with the system? You can find it here: https://github.com/markqvist/sideband

**Reply by githubques** [reply]
Yes, downloaded the newest version of SB for Linux. 
1) How was it designed (pure python on a .mu or .py file?)
2) Same question about the non-config sections of NomadNet.
3) Knowing that you have a heavy existing To-Do List, how difficult is it to update Reticulum/NomadNet/SB for non-Ubuntu Linux Distros? I understand Mobian is similar to Ubuntu with distros like Arch being different in structure.
4) Been doing some digging about .mu UI for when I've got some time to try building something small.
5) To send you a link to whatever .mu demo page I develop, is it as simple as sending an Announce or Network Request to your 14299468b08261697083940f54d13b02 LXMF address and wait for a reply?

**Reply by githubques** [reply]
One more note about SB, thought I had it installed but after typing "sideband" into terminal after reinstalling the latest version (sbapp-0.1.8) it doesn't load and instead gives this notice:

[INFO  ] [Logger] Record Log in ......
[INFO  ] [Kivy] v2.1.0
[INFO  ] [Kivy] Installed at ... 
[INFO  ] [Python] v3.10.4 (main, date info) [GCC 11.2.0]
[INFO  ] [Python] Interpreter at "user/bin/python3"
[INFO  ] [Python] Purge log fired. Processing...
[INFO  ] [Python] Purge finished!

**Reply by markqvist** [reply]
> 1. How was it designed (pure python on a .mu or .py file?)

Sideband is written in Python, and the user interface uses the Kivy and KivyMD libraries

> 2. Same question about the non-config sections of NomadNet.

NomadNet is also written in Python, but uses a library called Urwid to render the text-based user interface. Under the hood, Urwid uses the Curses library.

> 3. Knowing that you have a heavy existing To-Do List, how difficult is it to update Reticulum/NomadNet/SB for non-Ubuntu Linux Distros? I understand Mobian is similar to Ubuntu with distros like Arch being different in structure.

Reticulum already runs on all Linux distros, more or less. If Python 3 is available for the distro, it is almost certain that you can run Reticulum and NomadNet. I have run Reticulum and NomadNet even on embedded devices and stuff like e-book readers and similar.

The easiest way to install Reticulum on any distro is to use `pip` as outlined in the documentation, which will also set up all the commands, such as `rnsd`, `rnstatus` and the others. If `pip` is not available, it is still possible to install from source, but even on embedded devices and small systems like routers, it is often easy enough to install `pip`.

If there are situation where the `pip` install fails, please create an issue here on GitHub, so we can investigate how to make it work :)

> 4. Been doing some digging about .mu UI for when I've got some time to try building something small.
> 5. To send you a link to whatever .mu demo page I develop, is it as simple as sending an Announce or Network Request to your 14299468b08261697083940f54d13b02 LXMF address and wait for a reply?

Yes, you are welcome to send me a message with a link to your node and page to my LXMF address, I will be interested to see what you build :)

**Reply by markqvist** [reply]
Also, can you let me know what distro you are trying SB on where it fails to load?

**Reply by spaceinvadr** [reply]
Just wanted to chime in and say I’m having the exact same issue when attempting to run Sideband on Ubuntu 20.04, after installing using pip. It’s not clear to me where it’s failing as it never prints anything that looks like an error message. I’m getting the same terminal output as above.

**Reply by githubques** [reply]
Gave SB reply on "Sideband is now more useful #78" since it's OT for this thread and will better help others having the same issue.

1.&2. Thanks for the program listings. I went searching for .mu & micron and only found a python for kids tutorial and a corporation named micron (beyond your micron info under the Guide).

4.&.5 Sent the Ctrl-E ping to your network yesterday and didn't get a response unless I'm not looking in the right place.

**Reply by markqvist** [reply]
>1.&2. Thanks for the program listings. I went searching for .mu & micron and only found a python for kids tutorial and a corporation named micron (beyond your micron info under the Guide).

I don't think there is any information about micron available on the webpages or searchable via google or similar. The entire spec is in the guide and in the source code, and you will find more tips and examples on different nomadnet nodes and pages on the network.

>4.&.5 Sent the Ctrl-E ping to your network yesterday and didn't get a response unless I'm not looking in the right place.

I don't understand what you mean unfortunately. Is it the "Query network for keys" option in nomadnet you are talking about? It will only work if your own network is connected to a network that knows about the address you are trying to communicate with. If you are trying to send something to me, you could connect to one of the testnet hubs, which would in turn also connect you to networks that I am available on.

**Reply by githubques** [reply]
1&2. Good to know. Downloaded a bunch of scripts trying to get the codewithMu to work before coming across newer posts stating it can only be used with linux through a flashed iso on an external flash drive.

Having an issue getting kivy loaded. Is there a trusted source you recommend for install instructions? Jumped from one to the other with somewhat conflicting info leading me to probably download pieces in the wrong order or something since I can't get it to boot up despite showing at least some folders created. Some of the downloaded files from codewithMu & Kivy pinged as warnings when installing SB, but it works fine.

4&5. Should have been more specific, did I need to setup a testnet connection or was it ok as is. Looks like testnet was the answer which I've done and sent a short message to your LXMF address to see if it gets a response.

**Reply by markqvist** [reply]
>Downloaded a bunch of scripts trying to get the codewithMu to work before coming across newer posts stating it can only be used with linux through a flashed iso on an external flash drive.

I actually had no idea something called codewithMu existed. It seems to be a beginner-friendly python script editor. It has nothing to do with *micron* or `.mu` files that Nomadnet uses to format and render pages. It is just a coincidence that they have similar naming.

The whole idea with the *micron* format is that it is very light-weight, and can be edited and read in any plain-text editor, and can be viewed with formatting and colors in a program that supports rendering micron pages, like Nomadnet.

>sent a short message to your LXMF address to see if it gets a response.

I had your message waiting for me in my LXMF client when I returned to it. It seems like your client is no longer connected though, so the responses I have sent you should be waiting for you to download them from a propagation node.

**Comment by markqvist** [comment]
Thanks for the reports on failure to start up on Ubuntu. I have updated the wheel package, which was missing the correct reference to an internal module. You will have to uninstall Sideband, and reinstall from the repository (without using the local cached version). You can do this like so:

```
pip uninstall sbapp
pip install sbapp --no-cache-dir
```

That should make it work :)

**Reply by githubques** [reply]
Worked!

**Reply by markqvist** [reply]
Perfect!

**Comment by githubques** [comment]
Tried it on Ubuntu 22.04 & 20.04

I've going to repost under "Sideband is now more useful #78" since this 
is off topic and will better help others with the same issue.
